YAKL_atomics.h

```diff
@@ -102,9 +102,91 @@
         ::atomicMax( &update , value );
       }
     #endif
-  #endif
 
-  #ifdef __USE_HIP__
+  #elif defined(__USE_SYCL__)
+    template <typename T, sycl::access::address_space addressSpace =
+	      sycl::access::address_space::global_space>
+    using relaxed_atomic_ref =
+	sycl::ONEAPI::atomic_ref< T,
+				  sycl::ONEAPI::memory_order::relaxed,
+				  sycl::ONEAPI::memory_scope::device,
+				  addressSpace>;
+
+
+    template <typename T,
+	      sycl::access::address_space addressSpace =
+	      sycl::access::address_space::global_space>
+    __inline__ __attribute__((always_inline)) void atomicAdd(T &update , T value) {
+      relaxed_atomic_ref<T, addressSpace>( update ).fetch_add( value );
+    }
+
+    ////////////////////////////////////////////////////////////
+    // SYCL's atomics for reals could be quite slow with different Intel hardware
+    ////////////////////////////////////////////////////////////
+
+    template <typename T=float, sycl::access::address_space addressSpace =
+              sycl::access::address_space::global_space>
+    __inline__ __attribute__((always_inline)) void atomicAdd(float &addr , float operand) {
+      static_assert(sizeof(float) == sizeof(int), "Mismatched type size");
+
+      sycl::atomic<int, addressSpace> obj(
+        (sycl::multi_ptr<int, addressSpace>(reinterpret_cast<int *>(&addr))));
+
+      int old_value;
+      float old_float_value;
+
+      do {
+        old_value = obj.load(sycl::ONEAPI::memory_order::relaxed);
+        old_float_value = *reinterpret_cast<const float *>(&old_value);
+        const float new_float_value = old_float_value + operand;
+        const int new_value = *reinterpret_cast<const int *>(&new_float_value);
+        if (obj.compare_exchange_strong(old_value, new_value, sycl::ONEAPI::memory_order::relaxed))
+          break;
+      } while (true);
+
+      return;
+    }
+
+    template <typename T=double, sycl::access::address_space addressSpace =
```

Comment:
In the future, Intel GPUs support this, use hardware.

Also, need to either template or runtime differentiate Intel hardware to determine what supports this in hardware and what doesn't.